In [28]:
# import necessary libraries

import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
import seaborn as sns
import matplotlib.pyplot as plt

## Preliminary Data Inspection

This class is designed to streamline the initial exploration and quality assessment of a daraset. Providing methods like loading the datatset, inspecting the structure, previewing, identifying missing values and detecting duplicates.



In [6]:
class PreliminaryDataInspection():
    '''
    This class performs preliminary data inspection.
    '''

    def __init__(self, df=None, file_path=None, sep=","):
        '''
        Initializes the class by assigning a DataFrame or loading it from a file.

        Args:
            df (pd.DataFrame, optional): DataFrame to initialize the class with.
            file_path (str, optional): Path to the data file.
            sep (str, optional): Separator used in the file. Default is ",".
        '''
        if df is not None and isinstance(df, pd.DataFrame):
            self.df = df
            print("DataFrame provided directly.")
        elif file_path is not None:
            self.df = self.load_data(file_path, sep)
        else:
            self.df = None
            print("No valid DataFrame or file path provided.")

    def load_data(self, file_path, sep=","):
        '''
        Loads a data file and returns a DataFrame.

        Args:
            file_path (str): Path to the data file.
            sep (str): Separator used in the file. Default is ",".

        Returns:
            pd.DataFrame or None: The loaded DataFrame or None if file not found.
        '''
        try:
            df = pd.read_csv(file_path, sep=sep)
            print(f"Data successfully loaded from {file_path}.")
            return df
        except FileNotFoundError:
            print(f"File not found: {file_path}")
            return None
        except Exception as e:
            print(f"An error occurred: {e}")
            return None

    def check_info(self):
        '''
        Displays information about the DataFrame.
        '''
        if self.df is not None:
            print("DataFrame Info:")
            print(self.df.info())
        else:
            print("No DataFrame to inspect.")

    def check_head(self, n=5):
        '''
        Displays the first few rows of the DataFrame.

        Args:
            n (int): Number of rows to display. Default is 5.
        '''
        if self.df is not None:
            print(f"First {n} rows of the DataFrame:")
            print(self.df.head(n))
        else:
            print("No DataFrame to inspect.")

    def check_missing_values(self):
        '''
        Checks for missing values in the DataFrame.
        '''
        if self.df is not None:
            print("Missing Values in DataFrame:")
            missing_values = self.df.isnull().sum()
            print(missing_values[missing_values > 0])
        else:
            print("No DataFrame to inspect.")

    def check_duplicates(self):
        '''
        Checks for duplicate rows in the DataFrame.
        '''
        if self.df is not None:
            duplicates = self.df.duplicated().sum()
            print(f"Number of duplicate rows: {duplicates}")
        else:
            print("No DataFrame to inspect.")


### Alarm Response (`Alarmresponses.txt`)

In [4]:
alarm_response = PreliminaryDataInspection('data/Alarmresponses.txt', sep='\t')

alarm_response.check_info()
alarm_response.check_head()
alarm_response.check_missing_values()
alarm_response.check_duplicates()

Data successfully loaded from data/Alarmresponses.txt.
DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4124815 entries, 0 to 4124814
Data columns (total 8 columns):
 #   Column        Dtype 
---  ------        ----- 
 0   alarmid       int64 
 1   userid        int64 
 2   DeviceType    int64 
 3   devices       object
 4   Response      int64 
 5   ResponseDate  object
 6   latitude      object
 7   longitude     object
dtypes: int64(4), object(4)
memory usage: 251.8+ MB
None
First 5 rows of the DataFrame:
   alarmid               userid  DeviceType devices  Response ResponseDate  \
0    78100  2210044967363094894           4     Sms         0          NaN   
1    79629  2210044967363094894           4     Sms         0          NaN   
2    82395  2210044967363094894           4     Sms         0          NaN   
3    89313  2210044967363094894           4     Sms         0          NaN   
4    43918   738033312678202846           4     Sms         0          NaN   



### Data Summary:

Alarm reponses contains 4,124,815 entries with 8 columns (4214815 x 8 matrix), the columns are:

- alarmid: Unique identifier for each alarm event (integer).
- userid: Unique identifier for the user who triggered the alarm (integer).
- DeviceType: Type of device used to trigger the alarm (integer).
- devices: Description of the device used (object). 193 entries are missing values.
- Response: Indicates if a volunteer responded to the alarm (0 - no response, 1 - responded) (integer).
- ResponseDate: Date and time the volunteer responded (object). 3,603,938 entries are missing values.
- latitude: Latitude coordinate of the alarm location (object).
- longitude: Longitude coordinate of the alarm location (object).

**Data Insights:**

- A large portion of the data (over 87%) has missing values in the RespinseDate column. As at now, no conclusion could be made regarding the missing date values. Depending on its correlation with the target variable, I might result to dropping the column if it is negatively correlated with the Response column.

- There are missing values in the devices column, indicating some descriptions of the triggering devices are unavailable.

- The data seems suitable for analyzing volunteer response patterns, effectiveness, and identifying areas for improvement.


### Alarms (`Alarms.txt`)

In [5]:
alarms = PreliminaryDataInspection('data/Alarms.txt', sep='\t')

alarms.check_info()
alarms.check_head()
alarms.check_missing_values()
alarms.check_duplicates()

Data successfully loaded from data/Alarms.txt.
DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60329 entries, 0 to 60328
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   alarmid         60329 non-null  int64 
 1   alarmlocaltime  60329 non-null  object
 2   Latitude        60329 non-null  object
 3   Longitude       60329 non-null  object
 4   AlarmStatus     60329 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 2.3+ MB
None
First 5 rows of the DataFrame:
   alarmid           alarmlocaltime   Latitude Longitude  AlarmStatus
0    29977  2020-01-01 00:23:14.200   52,26168  5,603156            1
1    29979  2020-01-01 00:25:36.730  51,900643  4,484577            1
2    29980  2020-01-01 00:34:54.213   52,63205  4,726312            1
3    29982  2020-01-01 00:58:24.897  51,263437  4,045439            1
4    29983  2020-01-01 00:59:01.363  51,334741  6,129295            1
Missing Values 

### Data Summary:

The data is stored in a pandas DataFrame containing 60,329 entries. There are 5 columns in the DataFrame:
- alarmid: Unique identifier for each alarm event (integer).
- alarmlocaltime: Local time of the alarm event (object).
- Latitude: Latitude coordinate of the alarm location (object).
- Longitude: Longitude coordinate of the alarm location (object).
- AlarmStatus: Status of the alarm (integer, likely indicating active/inactive).

**Data Insights:**

- The data has no missing values.
- The alarmlocaltime column contains timestamp information, which can be useful for temporal analysis.
- The Latitude and Longitude columns provide geographic information, enabling spatial analysis.
- The AlarmStatus column can help identify active alarms and potentially correlate with response times or other factors.

### Volunteer availability



In [6]:
volunteer_availability = PreliminaryDataInspection('data/VolunteerAvailability.txt', sep='\t')

volunteer_availability.check_info()
volunteer_availability.check_head()
volunteer_availability.check_missing_values()
volunteer_availability.check_duplicates()

Data successfully loaded from data/VolunteerAvailability.txt.
DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2879387 entries, 0 to 2879386
Data columns (total 8 columns):
 #   Column                Dtype 
---  ------                ----- 
 0   userid                int64 
 1   FromHour              int64 
 2   FromMin               int64 
 3   ToHour                int64 
 4   ToMin                 int64 
 5   AvailableWeekDay      int64 
 6   AvailabilityLocation  int64 
 7   CreatedDate           object
dtypes: int64(7), object(1)
memory usage: 175.7+ MB
None
First 5 rows of the DataFrame:
                userid  FromHour  FromMin  ToHour  ToMin  AvailableWeekDay  \
0  2210044967363094894         0        0      23     59                 0   
1  2210044967363094894         0        0      23     59                 1   
2  2210044967363094894         0        0      23     59                 2   
3  2210044967363094894         0        0      23     59              

In [7]:
volunteers = PreliminaryDataInspection('data/Volunteers.txt', sep='\t')

volunteers.check_info()
volunteers.check_head()
volunteers.check_missing_values()
volunteers.check_duplicates()

Data successfully loaded from data/Volunteers.txt.
DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 246088 entries, 0 to 246087
Data columns (total 6 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   userid         246088 non-null  int64 
 1   DateOfBirth    246088 non-null  object
 2   Latitude       246088 non-null  object
 3   Longitude      246088 non-null  object
 4   WorkLatitude   246088 non-null  object
 5   WorkLongitude  246088 non-null  object
dtypes: int64(1), object(5)
memory usage: 11.3+ MB
None
First 5 rows of the DataFrame:
                userid              DateOfBirth          Latitude  \
0 -4402443911528580314  1973-10-16 00:00:00.000         53,239679   
1  4147689039375744256  1989-07-11 00:00:00.000        52,3007863   
2 -7843085838038396722  1997-07-01 00:00:00.000  52,3327560424805   
3 -1646437989330162528  1976-01-30 00:00:00.000  52,6543579101563   
4  5678998634590506361  1962-05-14 00:0

## Create a Unified Dataset

In [8]:
merged_data = pd.merge(alarm_response.df, alarms.df, on='alarmid', how='left')
merged_data = pd.merge(merged_data, volunteers.df, on='userid', how='left')
merged_data = pd.merge(merged_data, volunteer_availability.df, on='userid', how='left')
merged_data.head()

alarmid               userid  DeviceType devices  Response ResponseDate  \
0    78100  2210044967363094894           4     Sms         0          NaN   
1    78100  2210044967363094894           4     Sms         0          NaN   
2    78100  2210044967363094894           4     Sms         0          NaN   
3    78100  2210044967363094894           4     Sms         0          NaN   
4    78100  2210044967363094894           4     Sms         0          NaN   

           latitude         longitude           alarmlocaltime Latitude_x  \
0  51,9987411499023  4,77543687820435  2023-09-08 15:19:27.160  51,999415   
1  51,9987411499023  4,77543687820435  2023-09-08 15:19:27.160  51,999415   
2  51,9987411499023  4,77543687820435  2023-09-08 15:19:27.160  51,999415   
3  51,9987411499023  4,77543687820435  2023-09-08 15:19:27.160  51,999415   
4  51,9987411499023  4,77543687820435  2023-09-08 15:19:27.160  51,999415   

   ...       Longitude_y  WorkLatitude WorkLongitude FromHour FromMin ToHour  \
0  ...  4,77543687820435             0             0      0.0     0.0   23.0   
1  ...  4,77543687820435             0             0      0.0     0.0   23.0   
2  ...  4,77543687820435             0             0      0.0     0.0   23.0   
3  ...  4,77543687820435             0             0      0.0     0.0   23.0   
4  ...  4,77543687820435             0             0      0.0     0.0   23.0   

  ToMin  AvailableWeekDay  AvailabilityLocation              CreatedDate  
0  59.0               0.0                   0.0  2023-05-25 19:59:53.413  
1  59.0               1.0                   0.0  2023-05-25 19:59:53.413  
2  59.0               2.0                   0.0  2023-05-25 19:59:53.413  
3  59.0               3.0                   0.0  2023-05-25 19:59:53.413  
4  59.0               4.0                   0.0  2023-05-25 19:59:53.413  

[5 rows x 24 columns]

In [27]:
data = PreliminaryDataInspection(df=merged_data)

data.check_info()
data.check_head()
data.check_missing_values()
data.check_duplicates()

DataFrame provided directly.
DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48348212 entries, 0 to 48348211
Data columns (total 24 columns):
 #   Column                Dtype  
---  ------                -----  
 0   alarmid               int64  
 1   userid                int64  
 2   DeviceType            int64  
 3   devices               object 
 4   Response              int64  
 5   ResponseDate          object 
 6   latitude              object 
 7   longitude             object 
 8   alarmlocaltime        object 
 9   Latitude_x            object 
 10  Longitude_x           object 
 11  AlarmStatus           int64  
 12  DateOfBirth           object 
 13  Latitude_y            object 
 14  Longitude_y           object 
 15  WorkLatitude          object 
 16  WorkLongitude         object 
 17  FromHour              float64
 18  FromMin               float64
 19  ToHour                float64
 20  ToMin                 float64
 21  AvailableWeekDay      float64


## DataPreprocessor Class

In [31]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
import seaborn as sns
import matplotlib.pyplot as plt

class DataPreprocessor:
    """
    A class for handling data preprocessing tasks, including cleaning,
    feature engineering, and preparation for analysis or modeling.
    """
    
    def __init__(self, df=None, file_path=None, sep=","):
        """
        Initializes the DataPreprocessor class.

        Args:
            df (pd.DataFrame): A DataFrame provided directly. Defaults to None.
            file_path (str): Path to the data file (CSV). Defaults to None.
            sep (str): Separator used in the file. Defaults to ",".
        """
        if df is not None:
            if isinstance(df, pd.DataFrame):
                self.df = df
                print("DataFrame provided directly.")
            else:
                raise ValueError("Provided data is not a valid DataFrame.")
        elif file_path is not None:
            self.df = self.load_data(file_path, sep)
        else:
            raise ValueError("Either a DataFrame or file path must be provided.")

    def load_data(self, file_path, sep=","):
        """
        Loads a dataset from a file.

        Args:
            file_path (str): Path to the data file.
            sep (str): Separator used in the file. Defaults to ",".

        Returns:
            pd.DataFrame: Loaded DataFrame.
        """
        try:
            df = pd.read_csv(file_path, sep=sep)
            print(f"Data successfully loaded from {file_path}.")
            return df
        except FileNotFoundError:
            raise FileNotFoundError(f"File not found: {file_path}")
        except Exception as e:
            raise Exception(f"An error occurred while loading data: {e}")

    def handle_missing_values(self, threshold=50):
        """
        Handles missing values by dropping columns with high percentages
        of missing values and imputing remaining missing values.

        Args:
            threshold (float): Percentage threshold for dropping columns. Defaults to 50%.
        """
        missing_percentage = (self.df.isnull().sum() / len(self.df)) * 100
        print("Missing value percentages:\n", missing_percentage)

        # Drop columns with missing values above the threshold
        columns_to_drop = missing_percentage[missing_percentage > threshold].index
        self.df.drop(columns=columns_to_drop, axis=1, inplace=True)
        print(f"Columns dropped: {list(columns_to_drop)}")

        # Fill missing values in numeric columns with median
        numeric_cols = self.df.select_dtypes(include=['float64', 'int64']).columns
        self.df[numeric_cols] = self.df[numeric_cols].fillna(self.df[numeric_cols].median())

        # Fill missing values in categorical columns with "Unknown"
        categorical_cols = self.df.select_dtypes(include=['object']).columns
        self.df[categorical_cols] = self.df[categorical_cols].fillna("Unknown")

    def fix_data_types(self):
        """
        Fixes data types by converting latitude/longitude to numeric
        and converting date columns to datetime.
        """
        # Fix latitude and longitude columns
        lat_long_cols = ['latitude', 'longitude', 'Latitude_x', 'Longitude_x', 'Latitude_y', 'Longitude_y']
        for col in lat_long_cols:
            if col in self.df.columns:
                # Replace commas with dots and non-numeric values with NaN
                self.df[col] = (
                    self.df[col]
                    .replace(',', '.', regex=True)
                    .apply(pd.to_numeric, errors='coerce')
                )

        # Convert date columns to datetime
        date_cols = ['ResponseDate', 'alarmlocaltime', 'CreatedDate']
        for col in date_cols:
            if col in self.df.columns:
                self.df[col] = pd.to_datetime(self.df[col], errors='coerce')

    def remove_duplicates(self):
        """
        Removes duplicate rows from the dataset.
        """
        initial_rows = len(self.df)
        self.df.drop_duplicates(inplace=True)
        final_rows = len(self.df)
        print(f"Removed {initial_rows - final_rows} duplicate rows.")

    def scale_numeric_features(self, features):
        """
        Scales numeric features using StandardScaler.

        Args:
            features (list): List of numeric feature column names to scale.
        """
        scaler = StandardScaler()
        self.df[features] = scaler.fit_transform(self.df[features])

    def engineer_features(self):
        """
        Performs feature engineering, such as deriving response time
        and creating date-related features.
        """
        if 'ResponseDate' in self.df.columns and 'alarmlocaltime' in self.df.columns:
            self.df['response_time'] = (self.df['ResponseDate'] - self.df['alarmlocaltime']).dt.total_seconds() / 60

        # Add new features based on dates
        for col in ['ResponseDate', 'alarmlocaltime']:
            if col in self.df.columns:
                self.df[f"{col}_year"] = self.df[col].dt.year
                self.df[f"{col}_month"] = self.df[col].dt.month
                self.df[f"{col}_day"] = self.df[col].dt.day

    def visualize_data(self, feature):
        """
        Visualizes the distribution of a numeric feature.

        Args:
            feature (str): The feature to visualize.
        """
        if feature in self.df.columns:
            sns.histplot(self.df[feature], bins=50)
            plt.title(f"Distribution of {feature}")
            plt.show()
        else:
            print(f"Feature {feature} not found in the DataFrame.")

    def save_data(self, file_path):
        """
        Saves the cleaned DataFrame to a CSV file.

        Args:
            file_path (str): Path to save the file.
        """
        self.df.to_csv(file_path, index=False)
        print(f"Data saved to {file_path}.")


In [32]:
# Initialize the preprocessor with a DataFrame or file path
preprocessor = DataPreprocessor(df=data.df)

# Handle missing values
preprocessor.handle_missing_values(threshold=50)

# Fix data types
preprocessor.fix_data_types()

# Remove duplicates
preprocessor.remove_duplicates()

# Scale numeric features
numeric_features = ['FromHour', 'ToHour', 'FromMin', 'ToMin', 'AvailableWeekDay', 'AvailabilityLocation']
preprocessor.scale_numeric_features(numeric_features)

# Engineer new features
preprocessor.engineer_features()

# Visualize a feature
preprocessor.visualize_data('response_time')

# Save the cleaned data
preprocessor.save_data("data/cleaned_data.csv")


DataFrame provided directly.
Missing value percentages:
 alarmid                 0.0
userid                  0.0
DeviceType              0.0
devices                 0.0
Response                0.0
latitude                0.0
longitude               0.0
alarmlocaltime          0.0
Latitude_x              0.0
Longitude_x             0.0
AlarmStatus             0.0
DateOfBirth             0.0
Latitude_y              0.0
Longitude_y             0.0
WorkLatitude            0.0
WorkLongitude           0.0
FromHour                0.0
FromMin                 0.0
ToHour                  0.0
ToMin                   0.0
AvailableWeekDay        0.0
AvailabilityLocation    0.0
CreatedDate             0.0
dtype: float64
Columns dropped: []
Removed 0 duplicate rows.
Feature response_time not found in the DataFrame.
Data saved to data/cleaned_data.csv.


## Get Preprocessed Dataset and Sample 2,500 examples with no missing data from each class to perform further EDA due to size complexity

In [1]:
import pandas as pd

df = pd.read_csv('data/cleaned_data.csv')
df.head()

alarmid               userid  DeviceType devices  Response   latitude  \
0    78100  2210044967363094894           4     Sms         0  51.998741   
1    78100  2210044967363094894           4     Sms         0  51.998741   
2    78100  2210044967363094894           4     Sms         0  51.998741   
3    78100  2210044967363094894           4     Sms         0  51.998741   
4    78100  2210044967363094894           4     Sms         0  51.998741   

   longitude           alarmlocaltime  Latitude_x  Longitude_x  ...  FromHour  \
0   4.775437  2023-09-08 15:19:27.160   51.999415     4.774275  ... -0.360719   
1   4.775437  2023-09-08 15:19:27.160   51.999415     4.774275  ... -0.360719   
2   4.775437  2023-09-08 15:19:27.160   51.999415     4.774275  ... -0.360719   
3   4.775437  2023-09-08 15:19:27.160   51.999415     4.774275  ... -0.360719   
4   4.775437  2023-09-08 15:19:27.160   51.999415     4.774275  ... -0.360719   

    FromMin    ToHour     ToMin AvailableWeekDay AvailabilityLocation  \
0 -0.127003  0.482327  0.585447        -1.505984            -0.861165   
1 -0.127003  0.482327  0.585447        -1.004128            -0.861165   
2 -0.127003  0.482327  0.585447        -0.502271            -0.861165   
3 -0.127003  0.482327  0.585447        -0.000415            -0.861165   
4 -0.127003  0.482327  0.585447         0.501442            -0.861165   

               CreatedDate  alarmlocaltime_year  alarmlocaltime_month  \
0  2023-05-25 19:59:53.413                 2023                     9   
1  2023-05-25 19:59:53.413                 2023                     9   
2  2023-05-25 19:59:53.413                 2023                     9   
3  2023-05-25 19:59:53.413                 2023                     9   
4  2023-05-25 19:59:53.413                 2023                     9   

   alarmlocaltime_day  
0                   8  
1                   8  
2                   8  
3                   8  
4                   8  

[5 rows x 26 columns]

In [2]:
# Random Sample

def sample_from_dataframe(df, response_column='Response', sample_size=2500):
    """
    Randomly samples a fixed number of rows for each class in the Response column from a DataFrame.

    Parameters:
        df (pd.DataFrame): The loaded DataFrame.
        response_column (str): The column representing the classes.
        sample_size (int): The number of rows to sample for each class.

    Returns:
        pd.DataFrame: A smaller, balanced DataFrame.
    """
    # Ensure the Response column exists
    if response_column not in df.columns:
        raise ValueError(f"'{response_column}' column not found in the DataFrame.")
    
    # Sample data for each class
    sampled_data = (
        df.groupby(response_column)
          .apply(lambda x: x.sample(n=sample_size, random_state=42))
          .reset_index(drop=True)
    )
    
    return sampled_data

sampled_df = sample_from_dataframe(df)

print(sampled_df['Response'].value_counts())



Response
0    2500
1    2500
2    2500
Name: count, dtype: int64


/var/folders/lv/jcj3v7t92nv_xc98786nj4wm0000gn/T/ipykernel_75040/3351435095.py:22: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  .apply(lambda x: x.sample(n=sample_size, random_state=42))


In [3]:
sampled_df.head()

alarmid               userid  DeviceType devices  Response   latitude  \
0    71135   -43046965303985123           0     App         0  52.191235   
1    69199  2461588321794931671           0     App         0  52.365368   
2    84307  7233580007123251296           0     App         0  52.746689   
3    44619 -4183063865334114250           4     Sms         0  51.751053   
4    80944 -5543083286062164828           0     App         0  50.864712   

   longitude           alarmlocaltime  Latitude_x  Longitude_x  ...  FromHour  \
0   6.905455  2023-02-08 17:26:39.333   52.195694     6.896947  ... -0.360719   
1   4.853975  2022-12-23 11:17:26.360   52.362448     4.857253  ...  2.081725   
2   5.065581  2024-02-24 15:03:44.600   52.747531     5.067214  ... -0.360719   
3   5.266428  2021-02-14 17:10:24.520   51.750519     5.261850  ... -0.360719   
4   5.831041  2023-12-02 07:16:49.387   50.867397     5.833197  ... -0.360719   

    FromMin    ToHour     ToMin AvailableWeekDay AvailabilityLocation  \
0 -0.127003 -2.283653 -1.743488         1.505155             1.161217   
1 -0.127003  0.362067 -1.743488         1.003298            -0.861165   
2 -0.127003  0.482327  0.585447        -1.505984             1.161217   
3 -0.127003  0.482327  0.585447        -0.000415             1.161217   
4 -0.127003  0.482327  0.585447         1.003298             1.161217   

               CreatedDate  alarmlocaltime_year  alarmlocaltime_month  \
0  2022-01-10 13:42:17.323                 2023                     2   
1  2024-07-19 14:03:53.067                 2022                    12   
2  2021-04-29 20:17:29.170                 2024                     2   
3  2021-01-19 05:26:12.200                 2021                     2   
4  2023-10-11 15:00:04.763                 2023                    12   

   alarmlocaltime_day  
0                   8  
1                  23  
2                  24  
3                  14  
4                   2  

[5 rows x 26 columns]

In [4]:
sampled_df.info()
sampled_df.to_csv('data/sampled_data.csv', index=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500 entries, 0 to 7499
Data columns (total 26 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   alarmid               7500 non-null   int64  
 1   userid                7500 non-null   int64  
 2   DeviceType            7500 non-null   int64  
 3   devices               7500 non-null   object 
 4   Response              7500 non-null   int64  
 5   latitude              7500 non-null   float64
 6   longitude             7500 non-null   float64
 7   alarmlocaltime        7500 non-null   object 
 8   Latitude_x            7500 non-null   float64
 9   Longitude_x           7500 non-null   float64
 10  AlarmStatus           7500 non-null   int64  
 11  DateOfBirth           7500 non-null   object 
 12  Latitude_y            7480 non-null   float64
 13  Longitude_y           7480 non-null   float64
 14  WorkLatitude          7500 non-null   object 
 15  WorkLongitude        

In [ ]:
# Drop User and Alarm Id since they don't provide no meaningful data

sampled_df = df.drop(['alarmid', 'userid'], axis=1)

## Using sampled_data

In [167]:
sampled_data = PreliminaryDataInspection(file_path='data/sampled_data.csv', sep=',')


sampled_data.check_info()
sampled_data.check_head()
sampled_data.check_missing_values()
sampled_data.check_duplicates()

Data successfully loaded from data/sampled_data.csv.
DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500 entries, 0 to 7499
Data columns (total 26 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   alarmid               7500 non-null   int64  
 1   userid                7500 non-null   int64  
 2   DeviceType            7500 non-null   int64  
 3   devices               7500 non-null   object 
 4   Response              7500 non-null   int64  
 5   latitude              7500 non-null   float64
 6   longitude             7500 non-null   float64
 7   alarmlocaltime        7500 non-null   object 
 8   Latitude_x            7500 non-null   float64
 9   Longitude_x           7500 non-null   float64
 10  AlarmStatus           7500 non-null   int64  
 11  DateOfBirth           7500 non-null   object 
 12  Latitude_y            7480 non-null   float64
 13  Longitude_y           7480 non-null   float64
 14  Wor

According to `sampled.txt` I found the missing values in the following columns Latitude_y(20), Longitude_y(20), CreatedDate(20). Given this accounts for under 1%, I'll drop rows with missing data

In [168]:
data = sampled_data.df

data = data.dropna()

## Drop irrelevant columns using Domain Understanding

In [169]:
def drop_columns(df, columns):
    '''
    Drop columns
    
    Args:
        df: Dataframe
        columns: columns to drop
    
    Return
        Updated data frame
    '''
    if not isinstance(columns, list):
        raise TypeError(f'List expected')
    else:
        dropped = df.drop(columns=columns)
        
    return dropped

data = drop_columns(data, columns=['userid', 'DateOfBirth', 'alarmlocaltime', 'CreatedDate'])

In [170]:
data.head()

alarmid  DeviceType devices  Response   latitude  longitude  Latitude_x  \
0    71135           0     App         0  52.191235   6.905455   52.195694   
1    69199           0     App         0  52.365368   4.853975   52.362448   
2    84307           0     App         0  52.746689   5.065581   52.747531   
3    44619           4     Sms         0  51.751053   5.266428   51.750519   
4    80944           0     App         0  50.864712   5.831041   50.867397   

   Longitude_x  AlarmStatus  Latitude_y  ...  WorkLongitude  FromHour  \
0     6.896947            1   52.191235  ...              0 -0.360719   
1     4.857253            1   52.365368  ...              0  2.081725   
2     5.067214            1   52.377571  ...              0 -0.360719   
3     5.261850            1   51.751053  ...              0 -0.360719   
4     5.833197            1   50.868519  ...      5,8310409 -0.360719   

    FromMin    ToHour     ToMin  AvailableWeekDay  AvailabilityLocation  \
0 -0.127003 -2.283653 -1.743488          1.505155              1.161217   
1 -0.127003  0.362067 -1.743488          1.003298             -0.861165   
2 -0.127003  0.482327  0.585447         -1.505984              1.161217   
3 -0.127003  0.482327  0.585447         -0.000415              1.161217   
4 -0.127003  0.482327  0.585447          1.003298              1.161217   

   alarmlocaltime_year  alarmlocaltime_month  alarmlocaltime_day  
0                 2023                     2                   8  
1                 2022                    12                  23  
2                 2024                     2                  24  
3                 2021                     2                  14  
4                 2023                    12                   2  

[5 rows x 22 columns]

In [171]:
# Convert string values to numerical values

data['WorkLatitude'] = pd.to_numeric(data['WorkLatitude'], errors='coerce')
data['WorkLongitude'] = pd.to_numeric(data['WorkLongitude'], errors='coerce')


In [172]:
data.head()

alarmid  DeviceType devices  Response   latitude  longitude  Latitude_x  \
0    71135           0     App         0  52.191235   6.905455   52.195694   
1    69199           0     App         0  52.365368   4.853975   52.362448   
2    84307           0     App         0  52.746689   5.065581   52.747531   
3    44619           4     Sms         0  51.751053   5.266428   51.750519   
4    80944           0     App         0  50.864712   5.831041   50.867397   

   Longitude_x  AlarmStatus  Latitude_y  ...  WorkLongitude  FromHour  \
0     6.896947            1   52.191235  ...            0.0 -0.360719   
1     4.857253            1   52.365368  ...            0.0  2.081725   
2     5.067214            1   52.377571  ...            0.0 -0.360719   
3     5.261850            1   51.751053  ...            0.0 -0.360719   
4     5.833197            1   50.868519  ...            NaN -0.360719   

    FromMin    ToHour     ToMin  AvailableWeekDay  AvailabilityLocation  \
0 -0.127003 -2.283653 -1.743488          1.505155              1.161217   
1 -0.127003  0.362067 -1.743488          1.003298             -0.861165   
2 -0.127003  0.482327  0.585447         -1.505984              1.161217   
3 -0.127003  0.482327  0.585447         -0.000415              1.161217   
4 -0.127003  0.482327  0.585447          1.003298              1.161217   

   alarmlocaltime_year  alarmlocaltime_month  alarmlocaltime_day  
0                 2023                     2                   8  
1                 2022                    12                  23  
2                 2024                     2                  24  
3                 2021                     2                  14  
4                 2023                    12                   2  

[5 rows x 22 columns]

In [173]:
data = drop_columns(data, columns=['alarmlocaltime_year', 'alarmlocaltime_month', 'alarmlocaltime_day'])

df_alarm = pd.read_csv('data/Alarms.txt', sep='\t')
data = data.merge(df_alarm[['alarmid', 'alarmlocaltime']], on='alarmid', how='left')

data.head()


alarmid  DeviceType devices  Response   latitude  longitude  Latitude_x  \
0    71135           0     App         0  52.191235   6.905455   52.195694   
1    69199           0     App         0  52.365368   4.853975   52.362448   
2    84307           0     App         0  52.746689   5.065581   52.747531   
3    44619           4     Sms         0  51.751053   5.266428   51.750519   
4    80944           0     App         0  50.864712   5.831041   50.867397   

   Longitude_x  AlarmStatus  Latitude_y  Longitude_y  WorkLatitude  \
0     6.896947            1   52.191235     6.905455           0.0   
1     4.857253            1   52.365368     4.853975           0.0   
2     5.067214            1   52.377571     4.851784           0.0   
3     5.261850            1   51.751053     5.266428           0.0   
4     5.833197            1   50.868519     5.823439           NaN   

   WorkLongitude  FromHour   FromMin    ToHour     ToMin  AvailableWeekDay  \
0            0.0 -0.360719 -0.127003 -2.283653 -1.743488          1.505155   
1            0.0  2.081725 -0.127003  0.362067 -1.743488          1.003298   
2            0.0 -0.360719 -0.127003  0.482327  0.585447         -1.505984   
3            0.0 -0.360719 -0.127003  0.482327  0.585447         -0.000415   
4            NaN -0.360719 -0.127003  0.482327  0.585447          1.003298   

   AvailabilityLocation           alarmlocaltime  
0              1.161217  2023-02-08 17:26:39.333  
1             -0.861165  2022-12-23 11:17:26.360  
2              1.161217  2024-02-24 15:03:44.600  
3              1.161217  2021-02-14 17:10:24.520  
4              1.161217  2023-12-02 07:16:49.387

In [174]:
# Convert 'alarmlocaltime' to datetime
data['alarmlocaltime'] = pd.to_datetime(data['alarmlocaltime'])

# Define a function to categorize the time
def get_period_of_day(time):
    if time.hour < 6:
        return 'Night'
    elif time.hour < 12:
        return 'Morning'
    elif time.hour < 18:
        return 'Afternoon'
    else:
        return 'Evening'

# Apply the function to create a new column
data['period_of_day'] = data['alarmlocaltime'].dt.time.apply(get_period_of_day)

data.head()

alarmid  DeviceType devices  Response   latitude  longitude  Latitude_x  \
0    71135           0     App         0  52.191235   6.905455   52.195694   
1    69199           0     App         0  52.365368   4.853975   52.362448   
2    84307           0     App         0  52.746689   5.065581   52.747531   
3    44619           4     Sms         0  51.751053   5.266428   51.750519   
4    80944           0     App         0  50.864712   5.831041   50.867397   

   Longitude_x  AlarmStatus  Latitude_y  ...  WorkLatitude  WorkLongitude  \
0     6.896947            1   52.191235  ...           0.0            0.0   
1     4.857253            1   52.365368  ...           0.0            0.0   
2     5.067214            1   52.377571  ...           0.0            0.0   
3     5.261850            1   51.751053  ...           0.0            0.0   
4     5.833197            1   50.868519  ...           NaN            NaN   

   FromHour   FromMin    ToHour     ToMin  AvailableWeekDay  \
0 -0.360719 -0.127003 -2.283653 -1.743488          1.505155   
1  2.081725 -0.127003  0.362067 -1.743488          1.003298   
2 -0.360719 -0.127003  0.482327  0.585447         -1.505984   
3 -0.360719 -0.127003  0.482327  0.585447         -0.000415   
4 -0.360719 -0.127003  0.482327  0.585447          1.003298   

   AvailabilityLocation          alarmlocaltime period_of_day  
0              1.161217 2023-02-08 17:26:39.333     Afternoon  
1             -0.861165 2022-12-23 11:17:26.360       Morning  
2              1.161217 2024-02-24 15:03:44.600     Afternoon  
3              1.161217 2021-02-14 17:10:24.520     Afternoon  
4              1.161217 2023-12-02 07:16:49.387       Morning  

[5 rows x 21 columns]

In [175]:
data = drop_columns(data, columns=['alarmid', 'alarmlocaltime'])

data.head()

DeviceType devices  Response   latitude  longitude  Latitude_x  \
0           0     App         0  52.191235   6.905455   52.195694   
1           0     App         0  52.365368   4.853975   52.362448   
2           0     App         0  52.746689   5.065581   52.747531   
3           4     Sms         0  51.751053   5.266428   51.750519   
4           0     App         0  50.864712   5.831041   50.867397   

   Longitude_x  AlarmStatus  Latitude_y  Longitude_y  WorkLatitude  \
0     6.896947            1   52.191235     6.905455           0.0   
1     4.857253            1   52.365368     4.853975           0.0   
2     5.067214            1   52.377571     4.851784           0.0   
3     5.261850            1   51.751053     5.266428           0.0   
4     5.833197            1   50.868519     5.823439           NaN   

   WorkLongitude  FromHour   FromMin    ToHour     ToMin  AvailableWeekDay  \
0            0.0 -0.360719 -0.127003 -2.283653 -1.743488          1.505155   
1            0.0  2.081725 -0.127003  0.362067 -1.743488          1.003298   
2            0.0 -0.360719 -0.127003  0.482327  0.585447         -1.505984   
3            0.0 -0.360719 -0.127003  0.482327  0.585447         -0.000415   
4            NaN -0.360719 -0.127003  0.482327  0.585447          1.003298   

   AvailabilityLocation period_of_day  
0              1.161217     Afternoon  
1             -0.861165       Morning  
2              1.161217     Afternoon  
3              1.161217     Afternoon  
4              1.161217       Morning

# To categorical

In [176]:
import io

categorical_columns = ['DeviceType', 'devices', 'AlarmStatus', 'period_of_day']
data = pd.get_dummies(data, columns=categorical_columns, drop_first=False)

buffer = io.StringIO()
data.info(buf=buffer)
s = buffer.getvalue()

# Writing the output to a log file
with open('log.txt', 'w') as f:
    f.write(s)

In [177]:
# Convert all boolean columns to integers (0 or 1)
data[data.select_dtypes(include='bool').columns] = data.select_dtypes(include='bool').astype(int)

# Verify the change
data.head()

Response   latitude  longitude  Latitude_x  Longitude_x  Latitude_y  \
0         0  52.191235   6.905455   52.195694     6.896947   52.191235   
1         0  52.365368   4.853975   52.362448     4.857253   52.365368   
2         0  52.746689   5.065581   52.747531     5.067214   52.377571   
3         0  51.751053   5.266428   51.750519     5.261850   51.751053   
4         0  50.864712   5.831041   50.867397     5.833197   50.868519   

   Longitude_y  WorkLatitude  WorkLongitude  FromHour  ...  \
0     6.905455           0.0            0.0 -0.360719  ...   
1     4.853975           0.0            0.0  2.081725  ...   
2     4.851784           0.0            0.0 -0.360719  ...   
3     5.266428           0.0            0.0 -0.360719  ...   
4     5.823439           NaN            NaN -0.360719  ...   

   devices_Sms Pager Sms Pager  devices_Sms Sms  devices_Sms Sms App  \
0                            0                0                    0   
1                            0                0                    0   
2                            0                0                    0   
3                            0                0                    0   
4                            0                0                    0   

   devices_Unknown  AlarmStatus_1  AlarmStatus_3  period_of_day_Afternoon  \
0                0              1              0                        1   
1                0              1              0                        0   
2                0              1              0                        1   
3                0              1              0                        1   
4                0              1              0                        0   

   period_of_day_Evening  period_of_day_Morning  period_of_day_Night  
0                      0                      0                    0  
1                      0                      1                    0  
2                      0                      0                    0  
3                      0                      0                    0  
4                      0                      1                    0  

[5 rows x 43 columns]

In [178]:
import pandas as pd
import numpy as np

def haversine(lat1, lon1, lat2, lon2):
    # Convert latitude and longitude from degrees to radians
    lat1, lon1, lat2, lon2 = map(np.radians, [lat1, lon1, lat2, lon2])
    
    # Haversine formula
    dlat = lat2 - lat1
    dlon = lon2 - lon1
    a = np.sin(dlat / 2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon / 2)**2
    c = 2 * np.arcsin(np.sqrt(a))
    radius = 6371000  # Earth radius in meters (6371 km * 1000)
    distance = radius * c
    return distance


def calculate_distances(data):
    # Calculate distance between user and alarm
    data['distance_user_alarm'] = haversine(data['latitude'], data['longitude'], data['Latitude_x'], data['Longitude_x'])
    
    data['distance_user_volunteer'] = haversine(data['latitude'], data['longitude'], data['Latitude_y'], data['Longitude_y'])
    
    data['distance_user_work'] = haversine(data['latitude'], data['longitude'], data['WorkLatitude'], data['WorkLongitude'])
    
    data = data.drop(columns=['latitude', 'longitude', 'Latitude_x', 'Longitude_x', 'Latitude_y', 'Longitude_y', 'WorkLatitude', 'WorkLongitude'])
    
    return data

data = calculate_distances(data)
data.head()


Response  FromHour   FromMin    ToHour     ToMin  AvailableWeekDay  \
0         0 -0.360719 -0.127003 -2.283653 -1.743488          1.505155   
1         0  2.081725 -0.127003  0.362067 -1.743488          1.003298   
2         0 -0.360719 -0.127003  0.482327  0.585447         -1.505984   
3         0 -0.360719 -0.127003  0.482327  0.585447         -0.000415   
4         0 -0.360719 -0.127003  0.482327  0.585447          1.003298   

   AvailabilityLocation  DeviceType_0  DeviceType_3  DeviceType_4  ...  \
0              1.161217             1             0             0  ...   
1             -0.861165             1             0             0  ...   
2              1.161217             1             0             0  ...   
3              1.161217             0             0             1  ...   
4              1.161217             1             0             0  ...   

   devices_Unknown  AlarmStatus_1  AlarmStatus_3  period_of_day_Afternoon  \
0                0              1              0                        1   
1                0              1              0                        0   
2                0              1              0                        1   
3                0              1              0                        1   
4                0              1              0                        0   

   period_of_day_Evening  period_of_day_Morning  period_of_day_Night  \
0                      0                      0                    0   
1                      0                      1                    0   
2                      0                      0                    0   
3                      0                      0                    0   
4                      0                      1                    0   

   distance_user_alarm  distance_user_volunteer  distance_user_work  
0           762.996614                 0.000000        5.839183e+06  
1           393.633168                 0.000000        5.840363e+06  
2           144.372837             43513.925958        5.884067e+06  
3           320.679325                 0.000000        5.775628e+06  
4           334.703022               680.993244                 NaN  

[5 rows x 38 columns]

In [179]:
from sklearn.preprocessing import StandardScaler

response_column = 'Response'
numeric_features = ['FromHour', 'ToHour', 'FromMin', 'ToMin', 'AvailableWeekDay', 'AvailabilityLocation']

columns_to_scale = [col for col in data.columns if col not in [response_column] + numeric_features]

scaler = StandardScaler()
data_scaled = data.copy()

data_scaled[columns_to_scale] = scaler.fit_transform(data[columns_to_scale])

data_scaled.head()

Response  FromHour   FromMin    ToHour     ToMin  AvailableWeekDay  \
0         0 -0.360719 -0.127003 -2.283653 -1.743488          1.505155   
1         0  2.081725 -0.127003  0.362067 -1.743488          1.003298   
2         0 -0.360719 -0.127003  0.482327  0.585447         -1.505984   
3         0 -0.360719 -0.127003  0.482327  0.585447         -0.000415   
4         0 -0.360719 -0.127003  0.482327  0.585447          1.003298   

   AvailabilityLocation  DeviceType_0  DeviceType_3  DeviceType_4  ...  \
0              1.161217      0.711294     -0.016354     -0.710868  ...   
1             -0.861165      0.711294     -0.016354     -0.710868  ...   
2              1.161217      0.711294     -0.016354     -0.710868  ...   
3              1.161217     -1.405888     -0.016354      1.406731  ...   
4              1.161217      0.711294     -0.016354     -0.710868  ...   

   devices_Unknown  AlarmStatus_1  AlarmStatus_3  period_of_day_Afternoon  \
0        -0.011563       0.390794      -0.390794                 1.407575   
1        -0.011563       0.390794      -0.390794                -0.710442   
2        -0.011563       0.390794      -0.390794                 1.407575   
3        -0.011563       0.390794      -0.390794                 1.407575   
4        -0.011563       0.390794      -0.390794                -0.710442   

   period_of_day_Evening  period_of_day_Morning  period_of_day_Night  \
0              -0.637193              -0.633669            -0.313149   
1              -0.637193               1.578111            -0.313149   
2              -0.637193              -0.633669            -0.313149   
3              -0.637193              -0.633669            -0.313149   
4              -0.637193               1.578111            -0.313149   

   distance_user_alarm  distance_user_volunteer  distance_user_work  
0            -0.050429                -0.065736            0.149365  
1            -0.051699                -0.065736            0.154372  
2            -0.052555                 0.083679            0.339837  
3            -0.051949                -0.065736           -0.120340  
4            -0.051901                -0.063397                 NaN  

[5 rows x 38 columns]